### PR TITLE
[Move] Move to using Move for genesis, stdlib, and transaction scripts

### PR DESCRIPTION
**Pull request by tzakian**
This PR does a couple things:
* The base commit directly ports over the IR stdlib to Move. There is some cleanup and some Move-ifying of things to do, but I don't want to put that in this PR in order to keep this as "clean" and as similar to the original IR as possible. 
* This switches everything over to using the Move language (with the exception of tests); the genesis, transaction scripts, and stdlib modules are all now Move.

In the process of the switchover I wound up doing some cleanup around some funkiness in the transaction builder where we were parsing the program and then compiling it at different times, but this wasn't used or useful anywhere. We now parse and compile the transactions once, after that we just pull in the bytecode.

### CODE REVIEW

File: language/move-lang/stdlib/modules/libra_account.move [hunk 1]
```diff
@@ -4,11 +4,13 @@ address 0x0:
...
-    use 0x0::Event;
-    use 0x0::Transaction;
+    use 0x0::U64Util;
     use 0x0::AddressUtil;
+    use 0x0::BytearrayUtil;
+    use 0x0::LibraTransactionTimeout;
+    use 0x0::Transaction;
 
-    // Every Libra account has a LibraAccount::T resource
+    // Every Libra account has a LibraAccount.T resource
```
Comment (bob-wilson) [inline language/move-lang/stdlib/modules/libra_account.move]: This seems like it should be "LibraAccount::T"

Comment (bob-wilson) [reply language/move-lang/stdlib/modules/libra_account.move]: Several other places later in the file have the same thing, too.

File: language/move-lang/stdlib/modules/libra_account.move [hunk 2]
```diff
@@ -102,113 +150,203 @@ module LibraAccount {
...
-        // Abort if wWe already extracted the unique withdrawal capability for this account.
-        if (sender_account.delegated_withdrawal_capability) abort 11;
+        let sender = Transaction::sender();
+        let sender_account = borrow_global_mut<T>(sender);
+
+        // Abort if we already extracted the unique withdrawal capability for this account.
+        Transaction::assert(!sender_account.delegated_withdrawal_capability, 11);
 
-        // ensure uniqueness of the capability
+        // ensure the uniqueness of the capability
```
Comment (bob-wilson) [inline language/move-lang/stdlib/modules/libra_account.move]: Minor nit: This comment should be capitalized to match the style elsewhere in this file

File: language/move-lang/stdlib/modules/libra_account.move [hunk 3]
```diff
@@ -52,48 +62,86 @@ module LibraAccount {
...
+        // A globally unique ID for this event stream.
+        guid: bytearray,
+    }
 
     // Deposits the `to_deposit` coin into the `payee`'s account
     public fun deposit(payee: address, to_deposit: LibraCoin::T) acquires T {
-        // Check that the `to_deposit` coin is non-zero
-        let amount = LibraCoin::value(&to_deposit);
-        Transaction::assert(amount > 0, 7);
+        // Since we don't have bytearray literals in the source language at
```
Comment (bob-wilson) [inline language/move-lang/stdlib/modules/libra_account.move]: What's the plan for this? If we're going to add something in the language, it seems like this should have a "FIXME" comment or something. If we have no plans for that, and the u64_to_bytes is the right way to do it, then we should skip the comment here.

File: language/move-lang/stdlib/modules/libra_coin.move [hunk 4]
```diff
@@ -4,14 +4,19 @@ module LibraCoin {
...
 
-    // A singleton resource that grants access to `LibraCoin::mint`. Only the Association has one.
-    resource struct MintCapability {}
+    // A singleton resource that grants access to `LibraCoin.mint`. Only the Association has one.
+    resource struct MintCapability { }
 
-    // The sum of the values of all LibraCoin::T resources in the system
-    resource struct MarketCap { total_value: u64 }
+    resource struct MarketCap {
+        // The sum of the values of all LibraCoin.T resources in the system
```
Comment (bob-wilson) [inline language/move-lang/stdlib/modules/libra_coin.move]: Seems like it should be LibraCoin::T. Also later in this file....

File: language/move-lang/stdlib/modules/libra_coin.move [hunk 5]
```diff
@@ -4,14 +4,19 @@ module LibraCoin {
...
-    // The value of the coin. May be zero
-    resource struct T { value: u64 }
+    resource struct T {
+        // The value of the coin. May be zero
+        value: u64,
+    }
 
-    // A singleton resource that grants access to `LibraCoin::mint`. Only the Association has one.
-    resource struct MintCapability {}
+    // A singleton resource that grants access to `LibraCoin.mint`. Only the Association has one.
```
Comment (bob-wilson) [inline language/move-lang/stdlib/modules/libra_coin.move]: Seems like it should be LibraCoin::mint....

File: language/move-lang/stdlib/modules/libra_account.move (hunk 3 above)
Comment (tzakian) [reply language/move-lang/stdlib/modules/libra_account.move]: I think we should definitely have these. I'll add a FIXME.

File: language/move-lang/stdlib/modules/libra_account.move (hunk 1 above)
Comment (tzakian) [reply language/move-lang/stdlib/modules/libra_account.move]: Good spots. For some reason my regex didn't match these. I'll go through and fix them all up.

File: language/move-lang/stdlib/modules/libra_time.move [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+        // TODO: Should the initialized value be passed in to genesis?
+        let timer = CurrentTimeMicroseconds {microseconds: 0};
+        move_to_sender<CurrentTimeMicroseconds>(timer);
+    }
+
+    // Update the wall clock time by consensus. Requires VM privilege and will be invoked during block prologue.
+    public fun update_global_time(proposer: address, timestamp: u64) acquires CurrentTimeMicroseconds {
+        // Can only be invoked by LibraVM privilege.
+        // Fixme: uncomment when after adding support in functional test.
```
Comment (tnowacki) [inline language/move-lang/stdlib/modules/libra_time.move]: @tzakian is this fixme from you?

Comment (tzakian) [reply language/move-lang/stdlib/modules/libra_time.move]: No. This is an existing FIXME from IIRC @runtian-zhou

Comment (tnowacki) [reply language/move-lang/stdlib/modules/libra_time.move]: We should probably look into fixing this ASAP
@runtian-zhou @vgao1996

File: language/move-lang/stdlib/modules/libra_coin.move [hunk 7]
```diff
@@ -4,14 +4,19 @@ module LibraCoin {
...
     // A singleton resource that grants access to `LibraCoin::mint`. Only the Association has one.
-    resource struct MintCapability {}
+    resource struct MintCapability { }
 
-    // The sum of the values of all LibraCoin::T resources in the system
-    resource struct MarketCap { total_value: u64 }
+    resource struct MarketCap {
+        // The sum of the values of all LibraCoin::T resources in the system
+        //total_value: u128,
+        total_value: u64,
```
Comment (runtian-zhou) [inline language/move-lang/stdlib/modules/libra_coin.move]: FIXME?

File: language/move-lang/stdlib/modules/libra_system.move [hunk 8]
```diff
@@ -0,0 +1,650 @@
...
+        // FIXME: Update this once we have bytearray literals
+        id: U64Util::u64_to_bytes(0),
+        proposer: 0xA550C18,
+      });
+      LibraTimestamp::initialize_timer();
+    }
+
+    // ValidatorInfo public accessors
+
+    public fun get_validator_address(v: &ValidatorInfo): &address {
```
Comment (runtian-zhou) [inline language/move-lang/stdlib/modules/libra_system.move]: Should those functions return an owned primitive value rather than a reference?

File: language/move-lang/stdlib/modules/libra_system.move [hunk 9]
```diff
@@ -0,0 +1,650 @@
...
+    }
+
+    // Get the DiscoveryInfo for the ith validator
+    public fun get_ith_discovery_info(i: u64): DiscoveryInfo acquires DiscoverySet {
+        let discovery_vec_ref = &borrow_global<DiscoverySet>(0xD15C0).discovery_set;
+        Transaction::assert(i < Vector::length(discovery_vec_ref), 4);
+        *Vector::borrow(discovery_vec_ref, i)
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
```
Comment (runtian-zhou) [inline language/move-lang/stdlib/modules/libra_system.move]: Since we are porting the modules all together do you think we should move transaction fee to its own module?

File: language/move-lang/stdlib/modules/libra_time.move (hunk 6 above)
Comment (runtian-zhou) [reply language/move-lang/stdlib/modules/libra_time.move]: It was basically because we are calling into this function directly in functional test to test the behavior of block prologue, and adding this assertion will fail those tests. I'll add support for functional test soon so that we can add the assertions back.

File: language/move-lang/stdlib/modules/libra_system.move (hunk 9 above)
Comment (tzakian) [reply language/move-lang/stdlib/modules/libra_system.move]: I know we needed to include it in the system module before, but if we can separate things out that would be great!

However, I would be against doing this at the moment--I don't want to be introducing differences between this code and the IR modules until genesis generation has been swapped over. That way we should be able to generate with either of them. Once the switch is made splitting these makes sense if it's possible.

File: language/move-lang/stdlib/modules/gas_schedule.move [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+      storage: u64,
+    }
+
+    resource struct T {
+        instruction_schedule: Vector::T<Cost>,
+        native_schedule: Vector::T<Cost>,
+    }
+
+    // Initialize the table under the association account
+    fun initialize(gas_schedule: T) {
```
[the remark this answers is not in the document]
Comment (tzakian) [reply language/move-lang/stdlib/modules/gas_schedule.move]: I would need to think more about this. But at least at the moment I don't want to make that switch until we make the switch away from the IR genesis.

File: language/move-lang/stdlib/modules/libra_coin.move (hunk 7 above)
Comment (tzakian) [reply language/move-lang/stdlib/modules/libra_coin.move]: Yes. But soon not to be now that @tnowacki has landed them in the source language! I'll update now :)

File: language/move-lang/stdlib/modules/libra_account.move [hunk 11]
```diff
@@ -106,113 +151,208 @@ module LibraAccount {
...
-    public fun restore_key_rotation_capability(cap: Self::KeyRotationCapability) acquires T {
+    public fun restore_key_rotation_capability(cap: KeyRotationCapability) acquires T {
+        // Destroy the capability
         let KeyRotationCapability { account_address } = cap;
+        let account = borrow_global_mut<T>(account_address);
         // Update the flag for `account_address` to indicate that the capability has been restored.
         // The account owner will now be able to call rotate_authentication_key and
         // extract_sender_key_rotation_capability again
-        borrow_global_mut<T>(account_address).delegated_key_rotation_capability = false
+        account.delegated_key_rotation_capability = false;
```
[the remark this answers is not in the document]
Comment (tzakian) [reply language/move-lang/stdlib/modules/libra_account.move]: I think this is a property that should be encoded as a verifiable property. At the same time, having an assertion there as another statement of this fact shouldn't hurt. However, I'd prefer not to add it to this PR since I want to keep as much semantic similarity to the IR until we switch.

File: language/move-lang/stdlib/modules/libra_account.move [hunk 12]
```diff
@@ -52,52 +62,87 @@ module LibraAccount {
...
+        counter: u64,
     }
 
-    // Save an account to a given address if the address does not have an account resource yet
-    fun save_account(addr: address, account: Self::T) {
-        // FIXME this cant be native struct handle for Self::T is mismatched with
-        // the native dispatch
-        abort 0
+    // A handle for a event such that:
+    // 1. Other modules can emit event to this handle.
```
[the remark this answers is not in the document]
Comment (tzakian) [reply language/move-lang/stdlib/modules/libra_account.move]: Would it be "Other modules can emit events to through this handle instance"? @tnowacki ?

File: language/move-lang/stdlib/modules/libra_system.move (hunk 8 above)
Comment (tzakian) [reply language/move-lang/stdlib/modules/libra_system.move]: I think they should absolutely return an owned primitive value rather than a reference. But I'm very hesitant to make these changes until we switch (or at least, in a subsequent PR that changes both the IR and Move stdlib return types for these).

File: language/move-lang/stdlib/modules/gas_schedule.move (hunk 10 above)
[the remark this answers is not in the document]
Comment (runtian-zhou) [reply language/move-lang/stdlib/modules/gas_schedule.move]: "doing this will require changing the rust-level genesis generation code." -> as in what it is right now?

File: language/stdlib/modules/libra_account.move [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+        sender: address,
+        to_deposit: LibraCoin::T,
+        metadata: bytearray
+    ) acquires T {
+        // Check that the `to_deposit` coin is non-zero
+        let deposit_value = LibraCoin::value(&to_deposit);
+        Transaction::assert(deposit_value > 0, 7);
+
+        // Load the sender's account
+        let sender_account_ref = borrow_global_mut<T>(sender);
```
Comment (runtian-zhou) [inline language/stdlib/modules/libra_account.move]: This don’t need to be mut right?

Comment (tzakian) [reply language/stdlib/modules/libra_account.move]: It does doesn't it? In particular, we pass in the `sender_account_ref`'s `sent_events` field mutably in to `emit_event` just below this on line 125.

Comment (runtian-zhou) [reply language/stdlib/modules/libra_account.move]: Ahhh yes! You are right!

**Comment by dariorussi** [comment]
@vgao1996 you are probably aware of this already, just flagging to you given the bytearray -> vector changes coming

File: language/stdlib/modules/libra_coin.move [hunk 14]
```diff
@@ -29,43 +33,40 @@ module LibraCoin {
...
     // Create a new LibraCoin::T with a value of 0
     public fun zero(): T {
-        T { value: 0 }
+        T{value: 0}
     }
 
     // Public accessor for the value of a coin
-    public fun value(coin: &T): u64 {
-        coin.value
+    public fun value(coin_ref: &T): u64 {
```
Comment (dariorussi) [inline language/stdlib/modules/libra_coin.move]: should we have and insist on the `_ref` part in our code base? I just ask because we don't do it in, say, our rust code base and in general when coding in almost any language, but we push it here and there in our move code.
I am not too sure but it feels a bit verbose and diminishes readability in my mind.
This is more a question for everybody and for the future rather than a request for change

File: language/move-lang/stdlib/modules/libra_system.move (hunk 8 above)
Comment (dariorussi) [reply language/move-lang/stdlib/modules/libra_system.move]: I was doing a review of the stdlib modules a while ago and wondered about the same thing because I felt a bit "scared" on returning a reference ;) 
however most of those function are later on used with equality and thus returning a reference here is somewhat natural. It's also not a mutable reference.
I just mention because as I said I went to the same process as I assume @runtian-zhou did and I am still "scared" though it seems the right thing to do....
Those function being public is a bit more of a question I think, though they are used in testing so don't know... and yet if it was about tests driving a public API that would be disgusting...

File: language/move-lang/stdlib/modules/gas_schedule.move (hunk 10 above)
[the remark this answers is not in the document]
Comment (dariorussi) [reply language/move-lang/stdlib/modules/gas_schedule.move]: a note on the raw bytes: that would imply that we need and must specify a format for those bytes so that any VM, in any language can load them. It would also imply that a Move story for updates or reads would be impossible.
Not sure if that is a good thing or a bad thing ;) but figured I'd mention

File: language/stdlib/modules/libra_system.move [hunk 15]
```diff
@@ -0,0 +1,650 @@
...
+        let discovery_vec_ref = &mut discovery_set_ref.discovery_set;
+        let validator_set_changed = false;
+        let discovery_set_changed = false;
+
+        // (1) Check for additions
+        let i = 0;
+        let len = Vector::length(additions_ref);
+        // if additions is nonempty, we have an addition to process
+        if (len > 0) {
+            loop {
```
Comment (dariorussi) [inline language/stdlib/modules/libra_system.move]: I understand we want to keep things as close as to what they were for now but I wonder if we should consider rewriting all those loops with a `while` instead of an unconditional `loop` at some point.
It could make the code a bit more elegant and readable...
thoughts?

File: language/stdlib/modules/libra_coin.move (hunk 14 above)
Comment (tzakian) [reply language/stdlib/modules/libra_coin.move]: I personally am not a fan of this naming convention. In dynamic languages I can get behind them. But, in such a strongly-typed setting I don't think these names offer anything, and can in fact lead to issues (if e.g. the type is changed to no longer be a reference).

TL;DR: naming conventions (unlike the types) are not checked, we should rely on the checked stuff instead.

File: language/stdlib/modules/libra_system.move (hunk 15 above)
Comment (tzakian) [reply language/stdlib/modules/libra_system.move]: Agreed! The plan on this PR is to make the conversion stupidly straightforward. That way if there are any issues later on it should be easier to verify that the error didn't stem from a mis-port. There _will_ be a subsequent PR that Move-ifies these modules and fixes stuff like this.

**Comment by tzakian** [comment]
These changes have been merged in to #2705  and merged in to master. Closing.
